File: ios/SsdidWallet/Domain/Oid4vp/AuthorizationRequest.swift
import Foundation

struct AuthorizationRequest {
    let clientId: String
    let responseType: String?
    let responseMode: String?
    let responseUri: String?
    let nonce: String?
    let state: String?
    let presentationDefinition: String?
    let dcqlQuery: String?
    let requestUri: String?

    static func parse(_ uriString: String) -> Result<AuthorizationRequest, Error> {
        Result {
            guard let components = URLComponents(string: uriString) else {
                throw OpenId4VpError.missingClientId
            }

            let queryItems = components.queryItems ?? []

            func param(_ name: String) -> String? {
                queryItems.first(where: { $0.name == name })?.value
            }

            guard let clientId = param("client_id") else {
                throw OpenId4VpError.missingClientId
            }

            let requestUri = param("request_uri")

            // By-reference: only client_id and request_uri needed
            if let requestUri = requestUri {
                return AuthorizationRequest(
                    clientId: clientId,
                    responseType: nil,
                    responseMode: nil,
                    responseUri: nil,
                    nonce: nil,
                    state: nil,
                    presentationDefinition: nil,
                    dcqlQuery: nil,
                    requestUri: requestUri
                )
            }

            // By-value: validate all required parameters
            guard let responseType = param("response_type") else {
                throw OpenId4VpError.missingResponseType
            }
            guard responseType == "vp_token" else {
                throw OpenId4VpError.unsupportedResponseType(responseType)
            }

            guard let nonce = param("nonce") else {
                throw OpenId4VpError.missingNonce
            }

            let responseMode = param("response_mode") ?? "direct_post"
            guard responseMode == "direct_post" else {
                throw OpenId4VpError.unsupportedResponseMode(responseMode)
            }

            guard let responseUri = param("response_uri") else {
                throw OpenId4VpError.missingResponseUri
            }
            guard responseUri.hasPrefix("https://") else {
                throw OpenId4VpError.nonHttpsResponseUri(responseUri)
            }

            let presentationDefinition = param("presentation_definition")
            let dcqlQuery = param("dcql_query")

            if presentationDefinition != nil && dcqlQuery != nil {
                throw OpenId4VpError.ambiguousQuery
            }

            let state = param("state")

            // Validate client_id is HTTPS URL or DID
            try validateClientId(clientId)

            return AuthorizationRequest(
                clientId: clientId,
                responseType: responseType,
                responseMode: responseMode,
                responseUri: responseUri,
                nonce: nonce,
                state: state,
                presentationDefinition: presentationDefinition,
                dcqlQuery: dcqlQuery,
                requestUri: nil
            )
        }
    }

    private static func validateClientId(_ clientId: String) throws {
        let isHttps = clientId.hasPrefix("https://")
        let isDid = clientId.hasPrefix("did:")
        guard isHttps || isDid else {
            throw OpenId4VpError.invalidClientId(clientId)
        }
    }
}

File: android/app/src/main/java/my/ssdid/wallet/domain/oid4vp/AuthorizationRequest.kt
package my.ssdid.wallet.domain.oid4vp

import android.net.Uri
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

data class AuthorizationRequest(
    val clientId: String,
    val responseType: String? = null,
    val responseMode: String? = null,
    val responseUri: String? = null,
    val nonce: String? = null,
    val state: String? = null,
    val presentationDefinition: String? = null,
    val dcqlQuery: String? = null,
    val requestUri: String? = null
) {
    companion object {
        fun parse(uriString: String): Result<AuthorizationRequest> = runCatching {
            val uri = Uri.parse(uriString)

            val clientId = uri.getQueryParameter("client_id")
                ?: throw IllegalArgumentException("Missing required parameter: client_id")

            val requestUri = uri.getQueryParameter("request_uri")

            // By-reference: only client_id and request_uri needed
            if (requestUri != null) {
                require(requestUri.startsWith("https://")) { "request_uri must be HTTPS: $requestUri" }
                validateClientId(clientId)
                return@runCatching AuthorizationRequest(
                    clientId = clientId,
                    requestUri = requestUri
                )
            }

            // By-value: validate all required parameters
            val responseType = uri.getQueryParameter("response_type")
                ?: throw IllegalArgumentException("Missing required parameter: response_type")
            require(responseType == "vp_token") { "Unsupported response_type: $responseType" }

            val nonce = uri.getQueryParameter("nonce")
                ?: throw IllegalArgumentException("Missing required parameter: nonce")

            val responseMode = uri.getQueryParameter("response_mode") ?: "direct_post"
            require(responseMode == "direct_post") { "Unsupported response_mode: $responseMode" }

            val responseUri = uri.getQueryParameter("response_uri")
                ?: throw IllegalArgumentException("Missing required parameter: response_uri")
            require(responseUri.startsWith("https://")) {
                "response_uri must be HTTPS: $responseUri"
            }

            val presentationDefinition = uri.getQueryParameter("presentation_definition")
            val dcqlQuery = uri.getQueryParameter("dcql_query")

            if (presentationDefinition != null && dcqlQuery != null) {
                throw IllegalArgumentException(
                    "Request is ambiguous: both presentation_definition and dcql_query present"
                )
            }

            if (presentationDefinition == null && dcqlQuery == null) {
                throw IllegalArgumentException(
                    "Missing required query: neither presentation_definition nor dcql_query present"
                )
            }

            val state = uri.getQueryParameter("state")

            // Validate client_id is HTTPS URL or DID
            validateClientId(clientId)

            AuthorizationRequest(
                clientId = clientId,
                responseType = responseType,
                responseMode = responseMode,
                responseUri = responseUri,
                nonce = nonce,
                state = state,
                presentationDefinition = presentationDefinition,
                dcqlQuery = dcqlQuery
            )
        }

        fun parseJson(json: String): Result<AuthorizationRequest> = runCatching {
            val obj = Json.parseToJsonElement(json).jsonObject
            val clientId = obj["client_id"]?.jsonPrimitive?.content
                ?: throw IllegalArgumentException("Missing client_id in request object")

            validateClientId(clientId)

            val responseType = obj["response_type"]?.jsonPrimitive?.content
                ?: throw IllegalArgumentException("Missing response_type in request object")
            require(responseType == "vp_token") { "Unsupported response_type: $responseType" }

            val nonce = obj["nonce"]?.jsonPrimitive?.content
                ?: throw IllegalArgumentException("Missing nonce in request object")

            val responseMode = obj["response_mode"]?.jsonPrimitive?.content ?: "direct_post"
            require(responseMode == "direct_post") { "Unsupported response_mode: $responseMode" }

            val responseUri = obj["response_uri"]?.jsonPrimitive?.content
                ?: throw IllegalArgumentException("Missing response_uri in request object")
            require(responseUri.startsWith("https://")) {
                "response_uri must be HTTPS: $responseUri"
            }

            val presentationDefinition = obj["presentation_definition"]?.toString()
            val dcqlQuery = obj["dcql_query"]?.toString()

            if (presentationDefinition != null && dcqlQuery != null) {
                throw IllegalArgumentException(
                    "Request is ambiguous: both presentation_definition and dcql_query present"
                )
            }

            if (presentationDefinition == null && dcqlQuery == null) {
                throw IllegalArgumentException(
                    "Missing required query: neither presentation_definition nor dcql_query present"
                )
            }

            AuthorizationRequest(
                clientId = clientId,
                responseType = responseType,
                responseMode = responseMode,
                responseUri = responseUri,
                nonce = nonce,
                state = obj["state"]?.jsonPrimitive?.content,
                presentationDefinition = presentationDefinition,
                dcqlQuery = dcqlQuery
            )
        }

        private fun validateClientId(clientId: String) {
            val isHttps = clientId.startsWith("https://")
            val isDid = clientId.startsWith("did:")
            require(isHttps || isDid) { "client_id must be HTTPS URL or DID: $clientId" }
        }
    }
}
